adapters/missena/missena.go

```diff
@@ -1,82 +1,156 @@
+type UserParams struct {
+	Formats   []string       `json:"formats,omitempty"`
+	Placement string         `json:"placement,omitempty" default:"sticky"`
+	Sample    string         `json:"sample,omitempty"`
+	Settings  map[string]any `json:"settings,omitempty"`
 }
 
 type MissenaAdapter struct {
 	EndpointTemplate *template.Template
 }
 
+var defaultCur = "USD"
+
 // Builder builds a new instance of the Foo adapter for the given bidder with the given config.
 func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	endpoint, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
 	bidder := &adapter{
-		endpoint: config.Endpoint,
+		EndpointTemplate: endpoint,
 	}
 	return bidder, nil
 }
 
-func (a *adapter) makeRequest(missenaParams MissenaInternalParams, reqInfo *adapters.ExtraRequestInfo, impID string, request *openrtb2.BidRequest) (*adapters.RequestData, error) {
-	url := a.endpoint + "?t=" + missenaParams.ApiKey
+func getCurrency(currencies []string) (string, error) {
+	eurAvailable := false
+	for _, cur := range currencies {
+		if cur == defaultCur {
+			return defaultCur, nil
+		}
+		if cur == "EUR" {
+			eurAvailable = true
+		}
+	}
+	if eurAvailable {
+		return "EUR", nil
+	}
+	return "", fmt.Errorf("no currency supported %v", currencies)
+}
+
+func (a *adapter) getEndPoint(ext *openrtb_ext.ExtImpMissena) (string, error) {
+	endPointParams := macros.EndpointTemplateParams{
+		PublisherID: url.PathEscape(ext.APIKey),
+	}
+	return macros.ResolveMacros(a.EndpointTemplate, endPointParams)
+}
+
+func (a *adapter) makeRequest(imp openrtb2.Imp, request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo, params *openrtb_ext.ExtImpMissena, gdprApplies bool, consentString string) (*adapters.RequestData, error) {
+	url, err := a.getEndPoint(params)
+	if err != nil {
+		return nil, err
+	}
+	cur, err := getCurrency(request.Cur)
+	if err != nil {
+		cur = defaultCur
+	}
+
+	var floor float64
+	var floorCur string
+	if imp.BidFloor != 0 {
+		floor = imp.BidFloor
+		floorCur, err = getCurrency(request.Cur)
+		if err != nil {
+			floorCur = defaultCur
+			floor, err = requestInfo.ConvertCurrency(imp.BidFloor, imp.BidFloorCur, floorCur)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	var schain *openrtb2.SupplyChain
+	if request.Source != nil {
+		schain = request.Source.SChain
+	}
 
 	missenaRequest := MissenaAdRequest{
-		RequestId:        request.ID,
-		Timeout:          2000,
+		Adunit:           imp.ID,
+		COPPA:            request.Regs.COPPA,
+		Currency:         cur,
+		EIDs:             request.User.EIDs,
+		Floor:            floor,
+		FloorCurrency:    floorCur,
+		GDPR:             gdprApplies,
+		GDPRConsent:      consentString,
+		IdempotencyKey:   request.ID,
 		Referer:          request.Site.Page,
 		RefererCanonical: request.Site.Domain,
```

Comment:
Please add nil checks to avoid null pointer errors. Regs, User and Site are optional in OpenRTB request and PBS core doesn't guarantee they will be present in request.